### PR TITLE
Add more testing models for edge cases

### LINKED ISSUE #23: Add tests for database components and result values
The new functionality added in #22 shall be tested, e.g., different combinations of reading from and writing into database components, and different combinations of empty SEFFs and SEFFs with return values. The models can be simple but should test different combinations.

### PR DESCRIPTION
**Pull request by Entenwilli**
This PR adds more testing models to test certain scenarios occurring in the analysis:
It provides testing models and tests for
- [x] Functionality in two different assemblies (Handling different node characteristics in the same SEFF Component)
- [x] Simple database component model (Load and store dataflow variables in data stores)
- [x] Cyclic dependency in execution order of action sequences (Logger.error and Exception)
- [x] Warning for databases that are never written to (Logger.warn)
- [x] Simple example of returning variable (Availability of RETURN dataflow variable, returning dataflow variables from SEFFs)
- [x] TODO: Composite components
- [x] TODO: All sources for node characteristics, i.e., resource container, user, assembly context, (composite components?)

This PR closes #23 and relates to previous work in #3 and #5

### CODE REVIEW

**Comment by sebinside** [comment]
@Entenwilli I merged #22, please rebase this PR.

**Comment by Entenwilli** [comment]
All commits should be correct now and the history cleaned up. 

As of now the analysis does not support Composite Components, because the Sequence Finder is broken for Composite Components. As i have not looked in depth at the part of the Sequence Finder which finds Called SEFFs, it may require more time to fix it. See `PCMQueryUtils:134`

**Comment by Entenwilli** [comment]
Getting the label however should already be implemented, as all components that may have a label are present in the assembly context we create.
Only issue for now is, that we are unable to get the Effective Service Specification for the Called Component in Composite Components

**Comment by sebinside** [comment]
Thank you for the update. We should discuss how we proceed with composite components in the next meeting with @majuwa

**Comment by Entenwilli** [comment]
After some more digging around, i have found the issues in the analysis. I have fixed them, updated the tests and made sure everything still works after the modifications.

As this was the last TODO of this PR, this PR is now ready to be reviewed @sebinside, @majuwa

**Comment by sebinside** [comment]
Thank you for finishing this PR and the update. Maybe we can together have a look at the test cases in our call next year? Besides that, the work on the [first milestone](https://github.com/PalladioSimulator/Palladio-Addons-DataFlowConfidentiality-Analysis/milestone/1) seems to be completed. Congratulations 🎉

File: bundles/org.palladiosimulator.dataflow.confidentiality.analysis/src/org/palladiosimulator/dataflow/confidentiality/analysis/sequence/entity/pcm/PCMActionSequence.java [hunk 1]
```diff
@@ -147,20 +179,8 @@ public int compareTo(PCMActionSequence otherSequence) {
...
-			
-			if (requriredDatabases.containsAll(otherProvidedDatabases)) {
-				return 1;
-			} else if (otherRequiredDatabases.containsAll(providedDatabases)) {
-				return -1;
-			} else {
-				logger.error("Found incompatible set of Databases, Action Sequences depend on each other");
-				throw new IllegalStateException("Cylic loop of databases found in action sequences");
-			}
+			logger.error("Found incompatible set of Databases, Action Sequences depend on each other");
```
Comment (sebinside) [inline bundles/org.palladiosimulator.dataflow.confidentiality.analysis/src/org/palladiosimulator/dataflow/confidentiality/analysis/sequence/entity/pcm/PCMActionSequence.java:182]: The error message should also contain the information of the *cyclic* dependency

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/ListAppender.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.log4j.AppenderSkeleton;
+import org.apache.log4j.Level;
+import org.apache.log4j.spi.LoggingEvent;
+
+public class ListAppender extends AppenderSkeleton {
```
Comment (sebinside) [inline tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/ListAppender.java:11]: Please add a short comment why this class is needed at all (debugging logging and so on)

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/ConstraintResultTest.java [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+    	analysis.setLoggerLevel(Level.TRACE);
+    	List<ActionSequence> actionSequences = analysis.findAllSequences();
+    	List<ActionSequence> evaluatedSequences = analysis.evaluateDataFlows(actionSequences);
+    	
+    	List<AbstractActionSequenceElement<?>> results = evaluatedSequences.stream()
+    			.map(it -> analysis.queryDataFlow(it, constraint))
+    			.flatMap(it -> it.stream())
+    			.collect(Collectors.toList());
+    	
+    	assertEquals(constraintData.size(), results.size(), "Incorrect amount of violations found");
```
Comment (sebinside) [inline tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/ConstraintResultTest.java:166]: Here, but also at other parts of the code (e.g. in the data classes): The word `amount` usually specifies an uncountable number of items, while `count` refers to countable sets. See, e.g., this thread [here](https://ell.stackexchange.com/questions/3511/whats-the-difference-between-words-number-count-amount-and-quantity)

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/CharacteristicValueData.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package org.palladiosimulator.dataflow.confidentiality.analysis.constraint.data;
+
+public class CharacteristicValueData {
```
Comment (sebinside) [inline tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/CharacteristicValueData.java:3]: Could this be a record?

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/ConstraintData.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.Map;
+
+import org.palladiosimulator.dataflow.confidentiality.analysis.sequence.entity.AbstractActionSequenceElement;
+import org.palladiosimulator.dataflow.confidentiality.analysis.sequence.entity.CharacteristicValue;
+import org.palladiosimulator.dataflow.confidentiality.analysis.sequence.entity.DataFlowVariable;
+import org.palladiosimulator.dataflow.confidentiality.analysis.sequence.entity.pcm.AbstractPCMActionSequenceElement;
+
+import de.uka.ipd.sdq.identifier.Identifier;
+
+public class ConstraintData {
```
Comment (sebinside) [inline tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/ConstraintData.java:13]: Could this be a record?

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/ConstraintViolations.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
+package org.palladiosimulator.dataflow.confidentiality.analysis.constraint.data;
+
+import java.util.List;
+import java.util.Map;
+
+public class ConstraintViolations {
```
Comment (sebinside) [inline tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/ConstraintViolations.java:6]: Please add a private constructor as this seems to be an utility class

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/ConstraintResultTest.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    /**
+     * Provides Arguments for the testConstraintResults test.
+     * @return Returns a stream of Arguments containing the test model with constraint and expected results
+     */
+    private Stream<Arguments> provideTestConstraintResults() {
+    	Predicate<AbstractActionSequenceElement<?>> travelPlannerConstraint = node -> travelPlannerCondition(node);
+    	Predicate<AbstractActionSequenceElement<?>> internationalOnlineShopConstraint = node -> internationalOnlineShopCondition(node);
+
+    	StandalonePCMDataFlowConfidentialtyAnalysis mutipleAssembliesAnalysis = 
+    			super.initializeAnalysis(Paths.get("models", "MultipleAssembliesTest", "default.usagemodel"), Paths.get("models", "MultipleAssembliesTest", "default.allocation"));
```
Comment (sebinside) [inline tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/ConstraintResultTest.java:128]: I'm not happy with the name `MultipleAssembliesTest`. I understand, that you allocated multiple dependencies on a single resource, but the name indicates that there are somehow multiple assemblies in the model (or something like that). However, I do not have a good name either. Maybe something like `MultipleAssembliesOneResourceTest` to clarify it?

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/ConstraintData.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+	
+	public boolean hasNodeCharacteristic(CharacteristicValue actualCharacteristicValue) {
+		return true;
+	}
+	
+	public boolean hasDataFlowVariable(DataFlowVariable actualDataFlowVariable) {
+		return true;
+	}
+	
+	public int nodeCharacteristicsAmount() {
```
Comment (sebinside) [inline tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/ConstraintData.java:41]: Same comment about the word `amount` that should be replaced with `count`

**Review by sebinside — CHANGES_REQUESTED** [review]
Thank you for this PR, here is my first review. Unfortunately, I had problems opening the model files correctly, as expected due to #35. Still, I reviewed as much as possible. These are only minor remarks, after you updated the PR (and also please merge the main branch again), this can be merged. Thank you!

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/ConstraintData.java (hunk 5 above)
Comment (Entenwilli) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/data/ConstraintData.java:13]: I don't think it is necessary to use a record here, as the attributes of that class are only used internally

File: tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/ConstraintResultTest.java (hunk 7 above)
Comment (Entenwilli) [reply tests/org.palladiosimulator.dataflow.confidentiality.analysis.tests/src/org/palladiosimulator/dataflow/confidentiality/analysis/constraint/ConstraintResultTest.java:128]: I have renamed it to `OneAssemblyMutlipleResourceContainerTest` to indicate what this model is testing

**Review by sebinside — APPROVED** [review]
